### PR TITLE
Post and Put feature

### PR DESCRIPTION
**Pull request by serpient**
# Completes
<blockquote class="trello-card"><a href="https://trello.com/c/xsCMHped/41-25-http-server-can-handle-a-post-request-to-save-data">(2.5) HTTP Server can handle a POST request to save data</a></blockquote>
<blockquote class="trello-card"><a href="https://trello.com/c/P94uX4M2/42-25-http-server-can-handle-a-put-request-which-will-overwrite-an-existing-resources-data">(2.5) HTTP Server can handle a PUT request which will overwrite an existing resource&#39;s data</a></blockquote>

### CODE REVIEW

File: src/main/java/file_handler/FileHandler.java [hunk 1]
```diff
@@ -73,4 +84,54 @@ public static String getFileType(String path) {
         }
     }
 
+    public static void writeFile(String intendedFilePath, String fileType, byte[] fileContents) {
+        Path path = Paths.get(intendedFilePath + "." + fileType);
+
+        createDirectories(Paths.get(trimLastResource(intendedFilePath)));
+
+        try (OutputStream out = new BufferedOutputStream(
```
Comment (ssunday) [inline src/main/java/file_handler/FileHandler.java]: the spacing/line breaks seem a bit odd here

File: src/main/java/http_protocol/MIMETypes.java [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+
+    public static String getFileType(String mimeType) {
+        types.put("text/plain", "txt");
+        types.put("text/html", "html");
+        types.put("text/css", "css");
+        types.put("image/jpeg", "jpg");
+        types.put("image/png", "png");
+
+        return types.get(mimeType);
+
```
Comment (ssunday) [inline src/main/java/http_protocol/MIMETypes.java]: extra space

File: src/main/java/http_server/DirectoryCreator.java [hunk 3]
```diff
@@ -1,5 +1,6 @@
 package http_server;
 
+import file_handler.FileHandler;
```
Comment (ssunday) [inline src/main/java/http_server/DirectoryCreator.java]: curiously: is it snake case or usually camel case for the names?

Comment (serpient) [reply src/main/java/http_server/DirectoryCreator.java]: https://docs.oracle.com/javase/tutorial/java/package/namingpkgs.html

from the above link, i think my package should be `filehandler` and `httpserver` without underscores, but underscores are acceptable

Comment (ssunday) [reply src/main/java/http_server/DirectoryCreator.java]: interesting!

File: src/main/java/http_protocol/MIMETypes.java [hunk 4]
```diff
@@ -0,0 +1,18 @@
+package http_protocol;
+
+import java.util.HashMap;
+
+public class MIMETypes {
+    private static HashMap<String, String> types = new HashMap<>();
+
+    public static String getFileType(String mimeType) {
+        types.put("text/plain", "txt");
```
Comment (aaizuss) [inline src/main/java/http_protocol/MIMETypes.java]: Seems a bit odd that this function puts everything in the hash map and gets something out of it. Then every time you call `getFileType`, it's trying to add to the hash map before retrieving from the hash map.

File: src/main/java/http_protocol/RequestCreator.java [hunk 5]
```diff
@@ -1,77 +1,15 @@
 package http_protocol;
 
+import http_server.Parser;
 import http_server.Request;
 
-import java.util.HashMap;
-
 public class RequestCreator {
     public static Request from(String request) {
-        return new Request(parseMethod(request), parseRoute(request), parseBody(request), parseHeaders(request));
```
Comment (aaizuss) [inline src/main/java/http_protocol/RequestCreator.java]: I love seeing all of this red!

File: src/main/java/http_server/Parser.java [hunk 6]
```diff
@@ -0,0 +1,74 @@
+package http_server;
+
+import java.util.HashMap;
+
+public class Parser {
```
Comment (aaizuss) [inline src/main/java/http_server/Parser.java]: This name might be too generic - `RequestParser` might be more precise

File: src/test/java/http_server/DirectoryCreatorTest.java [hunk 7]
```diff
@@ -1,11 +1,25 @@
...
+import file_handler.FileHandler;
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import static org.junit.jupiter.api.Assertions.assertEquals;
 import java.util.List;
-import static org.junit.Assert.assertEquals;
 
 public class DirectoryCreatorTest {
+    @BeforeEach
```
Comment (aaizuss) [inline src/test/java/http_server/DirectoryCreatorTest.java]: Can you think of a way to test this functionality without creating real files in your file system?

File: src/main/java/http_server/Parser.java (hunk 6 above)
Comment (serpient) [reply src/main/java/http_server/Parser.java]: It actually parses Requests and Responses (parsing Responses are used in tests)

Comment (aaizuss) [reply src/main/java/http_server/Parser.java]: Interesting - I can't find where that's happening

Comment (serpient) [reply src/main/java/http_server/Parser.java]: Here you go: https://github.com/serpient/java_http_server/blob/fd8f09979480f8c52de4e3362b8b2779a1abf745/src/test/java/http_server/ServerTest.java

The file difference was too large for it to show

File: src/test/java/http_server/ServerTest.java [hunk 8]
```diff
@@ -1,169 +1,150 @@
...
-        String contentTypeHeader = "Content-Type: text/plain"  + Stringer.crlf;
-        String contentLengthHeader = "Content-Length: 48" + Stringer.crlf;
-        String body = Stringer.crlf + "Here are all my favorite movies:\n" + "- Harry Potter\n";
-        String response = responseLine + dateHeader + serverHeader + contentTypeHeader + contentLengthHeader + body;
-
         MockClientSocket mockClientSocket = new MockClientSocket(request);
         Session session = new Session(mockClientSocket, router);
-
         session.run();
+        String response = mockClientSocket.getSentData();
```
Comment (serpient) [inline src/test/java/http_server/ServerTest.java]: all the tests have the same setup (passed in a request, and run the session) -- extract this all into one function so the tests look cleaner

File: src/test/java/http_server/FileHandlerTest.java [hunk 9]
```diff
@@ -1,24 +1,22 @@
...
     @Test
     public void static_files_in_public_directory_are_rendered_as_a_HTML_page_with_contents_listed() {
-        FileHandler fileHandler = new FileHandler();
         String path = "./public";
 
-        assertEquals(true, fileHandler.readDirectoryContents(path).contains("Home.html"));
-        assertEquals(true, fileHandler.readDirectoryContents(path).contains("TurtleTab.txt"));
-        assertEquals(true, fileHandler.readDirectoryContents(path).contains("japan.png"));
-        assertEquals(true, fileHandler.readDirectoryContents(path).contains("water.png"));
+        assertEquals(true, FileHandler.readDirectoryContents(path).contains("Home.html"));
```
Comment (aaizuss) [inline src/test/java/http_server/FileHandlerTest.java]: These tests will break if you remove these files from the directory on your computer, right? Can you think of a way to test the file handler without relying on a real directory with these specific files existing on your computer?

Comment (serpient) [reply src/test/java/http_server/FileHandlerTest.java]: I'll put in a In-Memory repository!

**Comment by serpient** [comment]
@aaizuss  http_server_spec repo -> https://github.com/serpient/http_server_spec

**Pull request merged**